### PR TITLE
iox-#1394 fix axivion violation for type traits hpp types hpp

### PR DESCRIPTION
**Pull request by saif-at-github**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1394 (partly)

### CODE REVIEW

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp [hunk 1]
```diff
@@ -55,17 +55,17 @@ using add_const_conditionally_t = typename add_const_conditionally<T, C>::type;
 /// @endcode
 ///
 template <typename>
-constexpr bool always_false_v = false;
-
+constexpr bool always_false_v{false};
 // windows defines __cplusplus as 199711L
-#if __cplusplus < 201703L && !defined(_WIN32)
+// AXIVION DISABLE STYLE AutosarC++19_03-A16.0.1: TODO
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp:60]: The better approach would be to put this into the iceoryx platform and remove the axivion suppression.

In every platform under `iceoryx_hoofs/platform/{linux,mac,qnx,unix,win}/include/platform_settings.hpp` you find platform specific settings and adjustments. I would suggest to add there the following alias for windows, mac and unix
```cxx
template<typename C, typename... Cargs>
using invoke_result = std::invoke_result<C, Cargs...>;
```

For qnx and linux I would add:
```cxx
template <typename C, typename... Cargs>
using invoke_result = std::result_of<C(Cargs...)>;
```

With this we have to adjust the used `invoke_result` here to `platform::invoke_result`. Then we can remove the whole macro and have a very clean solution.

Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp]: @elfenpiff That's a very good suggested approach. have implemented it. Thanks.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp [hunk 2]
```diff
@@ -75,11 +67,13 @@ struct is_invocable
...
+    static constexpr std::true_type test(typename platform::invoke_result<C, As...>::type* f IOX_MAYBE_UNUSED) noexcept
     {
         return {};
     }
 
+    // AXIVION Next Construct AutosarC++19_03-A8.4.1 : we require a SFINEA failure case where all
+    // parameter types (non invokable ones) are allowed, this can be achieved with variadic arguments
     // This is chosen if Callable(ArgTypes) does not resolve to a valid type.
     template <typename C, typename... As>
     /// @NOLINTJUSTIFICATION we require a SFINEA failure case where all parameter types (non invokable ones) are
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp:79]: You can remove this and the following line. It's enough if the justification is in the Axivion line.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp [hunk 3]
```diff
@@ -105,12 +99,13 @@ struct is_invocable_r
...
+    test(std::enable_if_t<std::is_convertible<typename platform::invoke_result<C, As...>::type, ReturnType>::value>* f
              IOX_MAYBE_UNUSED) noexcept
     {
         return {};
     }
-
+    // AXIVION Next Construct AutosarC++19_03-A8.4.1 : we require a SFINEA failure case where all
+    // parameter types (non invokable ones) are allowed, this can be achieved with variadic arguments
     template <typename C, typename... As>
     /// @NOLINTJUSTIFICATION we require a SFINEA failure case where all parameter types (non invokable ones) are
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp:110]: Same here.

**Review by mossmaurice — COMMENTED** [review]
LGTM! Two small nitpicks.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp (hunk 2 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp]: @mossmaurice Sure

**Review by mossmaurice — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Pull request merged**